pkg/controller/master/virtualmachine/vm_volume_status_controller.go

```diff
@@ -0,0 +1,493 @@
+		return nil, err
+	}
+
+	var refOwnerVms []string
+	if refOwnerVms, err = h.getPVCRefOwnerVMs(pvc); err != nil {
+		return pvc, err
+	} else if len(refOwnerVms) == 0 {
+		// pvc is not attched to any owner yet
+		return pvc, nil
+	}
+
+	update := false
+	var volStatus string
+	var ok bool
+	// when pvc has !empty volume annotations, then update(add) to vm, otherwise remove from vm
+	if volStatus, ok = pvc.Annotations[util.AnnotationVolumeStatus]; ok && volStatus != "" {
+		update = true
+	}
+
+	for _, owner := range refOwnerVms {
+		namespace, vmName := ref.Parse(owner)
+		var err error
+		var vm *kubevirtapis.VirtualMachine
+		if vm, err = h.getVM(namespace, vmName); err != nil {
+			if !errors.IsNotFound(err) {
+				logrus.Infof("fail to get owner vm:(%s/%s), give up to update pvc status:%s, error:%s", namespace, vmName, volStatus, fmt.Errorf("%w", err))
+			}
+			continue
+		}
+
+		if update {
+			_, err = h.updateVolumeStatusToVM(vm, volumeStatusAnnotation{Name: pvc.Name, Status: volStatus})
+		} else {
+			_, err = h.removeVolumeStatusFromVM(vm, volumeStatusAnnotation{Name: pvc.Name})
+		}
+		if err != nil {
+			if !errors.IsConflict(err) {
+				logrus.Infof("pvc:(%s/%s) update status to owner vm:%s fail, error:%s", pvc.Namespace, pvc.Name, vm.Name, fmt.Errorf("%w", err))
+			}
+			return pvc, err
+		}
+	}
+
+	return pvc, nil
+}
+
+// when pvc is used by vm, it is not allowed to delte from webUI
+func (h *VMVolumeStatusController) OnPVCRemove(key string, pvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+	if pvc == nil {
+		return nil, nil
+	}
+
+	var refOwnerVms []string
+	var err error
+	if refOwnerVms, err = h.getPVCRefOwnerVMs(pvc); err != nil {
+		return nil, err
+	} else if len(refOwnerVms) == 0 {
+		// pvc is not attched to any owner yet
+		return nil, nil
+	}
+
+	// remove pvc status from vm
+	for _, owner := range refOwnerVms {
+		namespace, vmName := ref.Parse(owner)
+		var err error
+		var vm *kubevirtapis.VirtualMachine
+		if vm, err = h.getVM(namespace, vmName); err != nil {
+			if !errors.IsNotFound(err) {
+				logrus.Infof("fail to get owner vm:(%s/%s), give up to remove pvc:%s status, error:%s", namespace, vmName, pvc.Name, fmt.Errorf("%w", err))
+			}
+			continue
+		}
+
+		_, err = h.removeVolumeStatusFromVM(vm, volumeStatusAnnotation{Name: pvc.Name})
+		if err != nil {
+			if !errors.IsConflict(err) {
+				logrus.Infof("pvc:(%s/%s) remove status from owner vm:%s fail, error:%s", pvc.Namespace, pvc.Name, vmName, fmt.Errorf("%w", err))
+			}
+			return nil, err
+		}
+	}
+
+	return nil, nil
+}
+
+// after pvc is detached/unmounted from VM, remove its volume status from VM
+func (h *VMVolumeStatusController) removeVolumeStatusFromDetachedVM(pvc *corev1.PersistentVolumeClaim) (*corev1.PersistentVolumeClaim, error) {
+	var namespace, vmName string
+	var ok bool
+	if namespace, vmName, ok = h.decodeAnnotationDetachedVMString(pvc); !ok {
```

Comment:
Because namespace, vmName will be used later  in the function

There are 3 ways to do here:

A. try bester to use `:=` assignment and condition check in same line, like
if namespace, vmName, ok := h.decodeAnnotationDetachedVMString(pvc); !ok {
}
all following code will be nested in the {} as long as they need to use namespace or vmName, 
it is quite easy to write code which has several nested {}, for complex functionality, the multi nested {} will not be better than flat structure. 

B. separate variable assignment and condition check
namespace, vmName, ok := h.decodeAnnotationDetachedVMString(pvc)
if (!ok) {}

C. use `var ` to declare variable, then let assignment and condition check be in one line, as here

B,C are roughly same, they give the benefit of reduced level of nested {}
 and both are seen in Harvester project.
 
 @guangbochen are you talking about B or A ?

Reply:
Prefer to use `B` for better consistency, thanks.